### DISCUSSION TITLE
ternary-operator-in-bash

### DISCUSSION THREAD

**Discussion by giscus** [thomasnotfound/how.wtf#89] [Blog Posts]
# ternary-operator-in-bash

Many languages support a ternary operator: a short-form if/else statement for in-place conditional. This functionality can be mimicked in Bash.

https://how.wtf/ternary-operator-in-bash.html

**Comment by Whowants-Toknow** [comment]
Since Bash version 3.5 (may even have been earlier) it does have a ternary operator for 
integers and it works inside the arithmetic expression ((...)). 
See https://www.gnu.org/software/bash/manual/html_node/Shell-Arithmetic.html#Shell-Arithmetic

e.g.
    a=1
    b=2
    printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
a is greater
    a=3
    printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
a is not greater

Of course if not comparing integers then it's not available.

**Reply by Whowants-Toknow** [reply]
oops!! spot the deliberate mistake there? - just testing :)

**Reply by Whowants-Toknow** [reply]
a=1
b=2
printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
a is not greater
a=3
printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
a is greater

**Reply by thomasgtaylor** [reply]
Thanks for dropping this knowledge! Unfortunately, I can't replicate this behavior in Bash?

```bash
a=1
b=2
printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
a=3
printf "%s\n" $(( a>b ? "a is greater" : "a is not greater" ))
```

 I keep getting the following output:

```
script.sh: line 3: 1
: syntax error: invalid arithmetic operator (error token is "
")
script.sh: line 5: 3
: syntax error: invalid arithmetic operator (error token is "
")
```

**Reply by Whowants-Toknow** [reply]
Still got it wrong!! (never cut and paste when you are in a hurry!)
should have been;
a=1
b=2
printf "%s\n" $(( a>b ? 0 : 1 ))
1
a=3
printf "%s\n" $(( a>b ? 0 : 1 ))
0
have to remember this is arithmetic expansion so it wants to return a number ( I made this mistake myself when testing it out and then cut and pasted the wrong part of my history file - dooh!) So if dealing with string data Tom's approach is still better.

**Reply by thomasgtaylor** [reply]
Thank you for correcting it! This is awesome stuff - I did not realize it existed.
